apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc/states/decrypted_txs.rs

```diff
@@ -0,0 +1,51 @@
+use std::marker::PhantomData;
+
+use super::super::{thres, AllocStatus, BlockSpaceAllocator, TxBin};
+use super::{
+    BuildingDecryptedTxBatch, BuildingProtocolTxBatch, NextStateImpl, State,
+};
+
+impl State for BlockSpaceAllocator<BuildingDecryptedTxBatch> {
+    #[inline]
+    fn try_alloc(&mut self, tx: &[u8]) -> AllocStatus {
+        self.decrypted_txs.try_dump(tx)
+    }
+
+    #[inline]
+    fn try_alloc_batch<'tx, T>(&mut self, txs: T) -> AllocStatus
+    where
+        T: IntoIterator<Item = &'tx [u8]> + 'tx,
+    {
+        self.decrypted_txs.try_dump_all(txs)
+    }
+}
+
+impl NextStateImpl for BlockSpaceAllocator<BuildingDecryptedTxBatch> {
+    type Next = BlockSpaceAllocator<BuildingProtocolTxBatch>;
+
+    #[inline]
+    fn next_state_impl(mut self) -> Self::Next {
+        self.decrypted_txs.shrink();
+
+        // reserve space for protocol txs
+        let uninit = self.uninitialized_space_in_bytes();
+        self.protocol_txs = TxBin::init_over_ratio(uninit, thres::ONE_THIRD);
```

Comment:
It looks like we give protocol txs 1/3 of the remaining space after DKG decrypted txs have been allocated, but then DKG decrypted txs could take up the whole block potentially? Should we reserve space upfront for protocol txs instead?

https://github.com/anoma/namada/blob/dc86779f41b8a17d0d8734a911154b0123c585d4/apps/src/lib/node/ledger/shell/prepare_proposal/block_space_alloc.rs#L108-L120